File: typescript/0124-binary-tree-maximum-path-sum.ts
function maxPathSum(root: TreeNode | null): number {
    const res = [root.val];

    // return max path sum without split
    function dfs(root: TreeNode | null): number {
        if (!root) {
            return 0;
        }

        let leftMax = dfs(root.left);
        let rightMax = dfs(root.right);
        leftMax = Math.max(leftMax, 0);
        rightMax = Math.max(rightMax, 0);

        // compute max path sum WITH split
        res[0] = Math.max(res[0], root.val + leftMax + rightMax);
        return root.val + Math.max(leftMax, rightMax);
    }

    dfs(root);
    return res[0];
}

File: kotlin/0124-binary-tree-maximum-path-sum.kt
class Solution {
    fun maxPathSum(root: TreeNode?): Int {
        val res = IntArray(1) {Int.MIN_VALUE}
        maxPathSum(root, res)
        return res[0]
    }

    fun maxPathSum(root: TreeNode?, res: IntArray): Int {
        if (root == null)
            return 0

        val left = Math.max(0, maxPathSum(root.left, res))
        val right = Math.max(0, maxPathSum(root.right, res))
        res[0] = Math.max(res[0], root.`val` + left + right)

        return root.`val` + Math.max(left, right)
    }
}
